synthea-hiv/uploader/main.py

```diff
@@ -0,0 +1,156 @@
+    '--convert_to_openmrs',
+    action='store_true',
+    help='specify if uploading to OpenMRS')
+
+
+def list_all_files(directory: str) -> Dict[str, Set[pathlib.PosixPath]]:
+  """Lists JSON files under a directory.
+
+  Args:
+    directory: Directory containg JSON files
+
+  Returns:
+   Dictionary listing JSON files for hospitals, pracitioners, and patient
+   history
+  """
+
+  p = pathlib.Path(directory)
+  hospital_file = set(p.glob('hospitalInformation*.json'))
+  practitioner_file = set(p.glob('practitionerInformation*.json'))
+  patient_history_files = set(
+      p.glob('*.json')) - practitioner_file - hospital_file
+
+  file_type_dict = {
+      'hospital': hospital_file,
+      'practitioner': practitioner_file,
+      'patient_history': patient_history_files
+  }
+
+  logging.info("%s patients found in %s" % (len(patient_history_files), p.resolve()))
+  return file_type_dict
+
+
+def convert_to_bundle(
+    json_file: pathlib.PosixPath) -> bundle.Bundle:
+  """Loops through list of JSON files, and loads content of each file to create Bundle object.
+
+  Args:
+   json_file_set: Set containing JSON files that need to be read
+
+  Returns:
+    List containing Bundle objects
+  """
+
+  with open(json_file) as f:
+    data = json.loads(f.read())
+    return bundle.Bundle(json_file, data)
+
+
+def upload_openmrs(sink: fhir_hook.BaseHook,
+                   patient_bundle: bundle.Bundle):
+  """Upload Patient history bundles to OpenMRS.
+
+  For each bundle, we have to know the indiviual Patient, Encounters, and
+  Observations resources before uploading  as the OpenMRS FHIR Module does not
+  yet support uploading Bundles. OpenMRS does not suppport uploading the
+  hospital and practitioner files as well.
+
+
+  Args:
+    sink: OpenMRS FHIR endpoint to talk with
+    list_of_patients_bundles: list of all the Bundles that need to be uploaded
+  """
+
+  upload_handler = uploader.Uploader(sink)
+  patient_bundle.extract_resources()
+  upload_handler.upload_openmrs_bundle(patient_bundle)
+
+  # TODO(omarismail): implement save mapping
+  patient_bundle.save_mapping()
+
+
+def upload(sink: fhir_hook.BaseHook, each_bundle: bundle.Bundle):
+  """Upload bundles to a FHIR endpoint.
+
+  Server must be able to handle Bundle transactions.
+
+  Args:
+    sink: FHIR endpoint to talk with
+    file_dict: Dictionary listing JSON files for hospitals, pracitioners, and
+      patient history
+  """
+
+  upload_handler = uploader.Uploader(sink)
+  upload_handler.upload_bundle(each_bundle)
+
+
+def create_sink(url: str) -> fhir_hook.BaseHook:
+  if 'https://healthcare.googleapis.com' in url:
+    return fhir_hook.GcpHook(url)
+  else:
+    return fhir_hook.OpenMRSHook(url)
+
+
+if __name__ == '__main__':
+  logging.basicConfig(level=logging.INFO)
+  args = parser.parse_args()
+  json_file_dict = list_all_files(args.input_dir)
+  fhir_sink = create_sink(args.fhir_endpoint)
+
+  if args.convert_to_openmrs:
+    with multiprocessing.Pool() as p:
```

Comment:
Do we want to expose a parameter to control number of processes (which defaults to `os.cpu_count()`)?

Reply:
I did not want to change it as playing with it could slow down performance. I found the default value to be the "best"